### PR TITLE
store: return borrow to memtrie data on lookups

### PR DESCRIPTION
**Pull request by nagisa**
This is primarily intended to enable cheap existence lookups eventually but stylistically it seems quite awkward to force a conversion on the users of the API when a choice can be given at low-to-no cost.

### CODE REVIEW

**Comment by nagisa** [comment]
NB: this is intended to be a non-functional refactor. Any changes to coverage as thus are also bogus.

File: core/store/src/trie/mod.rs [hunk 1]
```diff
@@ -1378,7 +1380,9 @@ impl Trie {
         let charge_gas_for_trie_node_access =
             mode == KeyLookupMode::Trie || self.charge_gas_for_trie_node_access;
         if self.memtries.is_some() {
-            self.lookup_from_memory(key, charge_gas_for_trie_node_access)
+            self.lookup_from_memory(key, charge_gas_for_trie_node_access, |v| {
```
Comment (shreyan-gupta) [inline core/store/src/trie/mod.rs:1383]: Seems like this is the main change we are striving for?

**Review by shreyan-gupta — APPROVED** [review]

File: core/store/src/trie/mod.rs [hunk 2]
```diff
@@ -1252,11 +1253,12 @@ impl Trie {
     /// `charge_gas_for_trie_node_access` is used to control whether Trie node
     /// accesses incur any gas. Note that access to values is never charged here;
     /// it is only charged when the returned ref is dereferenced.
```
Comment (shreyan-gupta) [inline core/store/src/trie/mod.rs:1255]: nit: Could we update the description of this function to specify what `map_result` is and how is it used different to retrieve the value vs check existence?

File: core/store/src/trie/mod.rs (hunk 1 above)
Comment (nagisa) [reply core/store/src/trie/mod.rs:1383]: Umm, on one hand yes, on the other introducing a callback was forced by the fact that `self.memtries` are behind a lock. I would have rathered something cleaner TBH.

File: core/store/src/trie/mod.rs (hunk 2 above)
Comment (nagisa) [reply core/store/src/trie/mod.rs:1255]: Sure, can do. Although it feels somewhat weird to document internal helper functions to that level of detail, given that the code provides the same answer in a more concise, precise and formal way.

Comment (shreyan-gupta) [reply core/store/src/trie/mod.rs:1255]: Upto you, no worries!

Comment (nagisa) [reply core/store/src/trie/mod.rs:1255]: I think I managed to strike a good balance in the most recent commit.

**Pull request merged**